File: repotoire/detectors/health_delta.py
"""Health score delta calculator for estimating fix impact.

This module provides utilities for estimating how resolving a finding
would impact the overall health score. This enables before/after
comparisons when users review proposed fixes.

Example:
    ```python
    calculator = HealthScoreDeltaCalculator()
    delta = calculator.calculate_delta(current_metrics, finding)
    print(f"Fixing this would improve score by {delta.score_delta:.1f} points")
    ```
"""

from copy import deepcopy
from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional, Tuple

from repotoire.models import Finding, Severity, MetricsBreakdown


class ImpactLevel(str, Enum):
    """Impact level classification for health score changes."""

    CRITICAL = "critical"  # >5 points improvement or grade change
    HIGH = "high"  # 2-5 points improvement
    MEDIUM = "medium"  # 0.5-2 points improvement
    LOW = "low"  # <0.5 points improvement
    NEGLIGIBLE = "negligible"  # <0.1 points improvement


@dataclass
class HealthScoreDelta:
    """Result of a health score delta calculation.

    Attributes:
        before_score: Current overall health score
        after_score: Projected score after fix
        score_delta: Points improvement (positive = better)
        before_grade: Current letter grade
        after_grade: Projected letter grade after fix
        grade_improved: Whether grade would improve
        structure_delta: Points change in structure category
        quality_delta: Points change in quality category
        architecture_delta: Points change in architecture category
        impact_level: Classification of impact (low/medium/high/critical)
        affected_metric: Which metric would be improved
        finding_id: ID of the finding this delta relates to
        finding_severity: Severity of the finding
    """

    before_score: float
    after_score: float
    score_delta: float
    before_grade: str
    after_grade: str
    grade_improved: bool
    structure_delta: float
    quality_delta: float
    architecture_delta: float
    impact_level: ImpactLevel
    affected_metric: str
    finding_id: Optional[str] = None
    finding_severity: Optional[Severity] = None

    @property
    def grade_change_str(self) -> Optional[str]:
        """Return grade change as string (e.g., 'B → A') or None if unchanged."""
        if self.grade_improved:
            return f"{self.before_grade} → {self.after_grade}"
        return None

    def to_dict(self) -> Dict:
        """Convert to dictionary for JSON serialization."""
        return {
            "before_score": round(self.before_score, 1),
            "after_score": round(self.after_score, 1),
            "score_delta": round(self.score_delta, 2),
            "before_grade": self.before_grade,
            "after_grade": self.after_grade,
            "grade_improved": self.grade_improved,
            "grade_change": self.grade_change_str,
            "structure_delta": round(self.structure_delta, 2),
            "quality_delta": round(self.quality_delta, 2),
            "architecture_delta": round(self.architecture_delta, 2),
            "impact_level": self.impact_level.value,
            "affected_metric": self.affected_metric,
            "finding_id": self.finding_id,
            "finding_severity": self.finding_severity.value if self.finding_severity else None,
        }


@dataclass
class BatchHealthScoreDelta:
    """Result of calculating delta for multiple findings."""

    before_score: float
    after_score: float
    score_delta: float
    before_grade: str
    after_grade: str
    grade_improved: bool
    findings_count: int
    individual_deltas: List[HealthScoreDelta] = field(default_factory=list)

    def to_dict(self) -> Dict:
        """Convert to dictionary for JSON serialization."""
        return {
            "before_score": round(self.before_score, 1),
            "after_score": round(self.after_score, 1),
            "score_delta": round(self.score_delta, 2),
            "before_grade": self.before_grade,
            "after_grade": self.after_grade,
            "grade_improved": self.grade_improved,
            "grade_change": f"{self.before_grade} → {self.after_grade}" if self.grade_improved else None,
            "findings_count": self.findings_count,
            "individual_deltas": [d.to_dict() for d in self.individual_deltas],
        }


# Mapping from detector names to the metrics they affect
DETECTOR_METRIC_MAPPING: Dict[str, Tuple[str, str]] = {
    # (metric_name, category)
    "CircularDependencyDetector": ("circular_dependencies", "structure"),
    "GodClassDetector": ("god_class_count", "quality"),
    "DeadCodeDetector": ("dead_code_percentage", "quality"),
    "VultureDetector": ("dead_code_percentage", "quality"),
    "ArchitecturalBottleneckDetector": ("bottleneck_count", "structure"),
    "JscpdDetector": ("duplication_percentage", "quality"),
    "DuplicateRustDetector": ("duplication_percentage", "quality"),
    "LayerViolationDetector": ("layer_violations", "architecture"),
    "BoundaryViolationDetector": ("boundary_violations", "architecture"),
    "ModuleCohesionDetector": ("modularity", "structure"),
    "InappropriateIntimacyDetector": ("avg_coupling", "structure"),
    "FeatureEnvyDetector": ("avg_coupling", "structure"),
    "ShotgunSurgeryDetector": ("avg_coupling", "structure"),
    "MiddleManDetector": ("bottleneck_count", "structure"),
    "DataClumpsDetector": ("avg_coupling", "structure"),
}

# Grade thresholds (same as in engine.py)
GRADES: Dict[str, Tuple[float, float]] = {
    "A": (90, 100),
    "B": (80, 90),
    "C": (70, 80),
    "D": (60, 70),
    "F": (0, 60),
}


class HealthScoreDeltaCalculator:
    """Calculate health score deltas for individual or batched findings.

    This calculator estimates how resolving findings would impact the
    health score, enabling before/after comparisons in the UI.
    """

    # Category weights (must match engine.py)
    STRUCTURE_WEIGHT = 0.40
    QUALITY_WEIGHT = 0.30
    ARCHITECTURE_WEIGHT = 0.30

    def calculate_delta(
        self,
        metrics: MetricsBreakdown,
        finding: Finding,
    ) -> HealthScoreDelta:
        """Calculate health score delta for resolving a single finding.

        Args:
            metrics: Current metrics breakdown
            finding: The finding to simulate resolving

        Returns:
            HealthScoreDelta with before/after comparison
        """
        # Calculate current scores
        current_structure = self._score_structure(metrics)
        current_quality = self._score_quality(metrics)
        current_architecture = self._score_architecture(metrics)
        current_overall = self._calculate_overall(
            current_structure, current_quality, current_architecture
        )
        current_grade = self._score_to_grade(current_overall)

        # Simulate removing the finding's impact
        modified_metrics = self._remove_finding_impact(metrics, finding)

        # Calculate new scores
        new_structure = self._score_structure(modified_metrics)
        new_quality = self._score_quality(modified_metrics)
        new_architecture = self._score_architecture(modified_metrics)
        new_overall = self._calculate_overall(
            new_structure, new_quality, new_architecture
        )
        new_grade = self._score_to_grade(new_overall)

        # Calculate deltas
        score_delta = new_overall - current_overall
        structure_delta = new_structure - current_structure
        quality_delta = new_quality - current_quality
        architecture_delta = new_architecture - current_architecture

        # Determine affected metric
        affected_metric = self._get_affected_metric(finding.detector)

        # Classify impact level
        impact_level = self._classify_impact(
            score_delta, current_grade != new_grade
        )

        return HealthScoreDelta(
            before_score=current_overall,
            after_score=new_overall,
            score_delta=score_delta,
            before_grade=current_grade,
            after_grade=new_grade,
            grade_improved=new_grade < current_grade,  # A < B < C < D < F
            structure_delta=structure_delta,
            quality_delta=quality_delta,
            architecture_delta=architecture_delta,
            impact_level=impact_level,
            affected_metric=affected_metric,
            finding_id=finding.id if hasattr(finding, "id") else None,
            finding_severity=finding.severity,
        )

    def calculate_batch_delta(
        self,
        metrics: MetricsBreakdown,
        findings: List[Finding],
    ) -> BatchHealthScoreDelta:
        """Calculate health score delta for resolving multiple findings.

        Args:
            metrics: Current metrics breakdown
            findings: List of findings to simulate resolving

        Returns:
            BatchHealthScoreDelta with aggregate before/after comparison
        """
        if not findings:
            current_overall = self._calculate_overall(
                self._score_structure(metrics),
                self._score_quality(metrics),
                self._score_architecture(metrics),
            )
            current_grade = self._score_to_grade(current_overall)
            return BatchHealthScoreDelta(
                before_score=current_overall,
                after_score=current_overall,
                score_delta=0.0,
                before_grade=current_grade,
                after_grade=current_grade,
                grade_improved=False,
                findings_count=0,
            )

        # Calculate current scores
        current_structure = self._score_structure(metrics)
        current_quality = self._score_quality(metrics)
        current_architecture = self._score_architecture(metrics)
        current_overall = self._calculate_overall(
            current_structure, current_quality, current_architecture
        )
        current_grade = self._score_to_grade(current_overall)

        # Calculate individual deltas
        individual_deltas = []
        for finding in findings:
            delta = self.calculate_delta(metrics, finding)
            individual_deltas.append(delta)

        # Simulate removing all findings' impacts
        modified_metrics = deepcopy(metrics)
        for finding in findings:
            modified_metrics = self._remove_finding_impact(modified_metrics, finding)

        # Calculate new aggregate scores
        new_structure = self._score_structure(modified_metrics)
        new_quality = self._score_quality(modified_metrics)
        new_architecture = self._score_architecture(modified_metrics)
        new_overall = self._calculate_overall(
            new_structure, new_quality, new_architecture
        )
        new_grade = self._score_to_grade(new_overall)

        return BatchHealthScoreDelta(
            before_score=current_overall,
            after_score=new_overall,
            score_delta=new_overall - current_overall,
            before_grade=current_grade,
            after_grade=new_grade,
            grade_improved=new_grade < current_grade,
            findings_count=len(findings),
            individual_deltas=individual_deltas,
        )

    def _remove_finding_impact(
        self,
        metrics: MetricsBreakdown,
        finding: Finding,
    ) -> MetricsBreakdown:
        """Create modified metrics by removing one finding's contribution.

        Args:
            metrics: Current metrics
            finding: Finding to remove

        Returns:
            Modified MetricsBreakdown with finding's impact removed
        """
        modified = deepcopy(metrics)
        detector = finding.detector

        # Apply detector-specific adjustments
        if detector == "CircularDependencyDetector":
            modified.circular_dependencies = max(0, modified.circular_dependencies - 1)

        elif detector == "GodClassDetector":
            modified.god_class_count = max(0, modified.god_class_count - 1)

        elif detector in ("DeadCodeDetector", "VultureDetector"):
            # Estimate one dead code item
            total_nodes = modified.total_classes + modified.total_functions
            if total_nodes > 0:
                per_item_pct = 1.0 / total_nodes
                modified.dead_code_percentage = max(
                    0.0, modified.dead_code_percentage - per_item_pct
                )

        elif detector == "ArchitecturalBottleneckDetector":
            modified.bottleneck_count = max(0, modified.bottleneck_count - 1)

        elif detector in ("JscpdDetector", "DuplicateRustDetector"):
            # Estimate 0.5% reduction per duplicate finding
            modified.duplication_percentage = max(
                0.0, modified.duplication_percentage - 0.005
            )

        elif detector == "LayerViolationDetector":
            modified.layer_violations = max(0, modified.layer_violations - 1)

        elif detector == "BoundaryViolationDetector":
            modified.boundary_violations = max(0, modified.boundary_violations - 1)

        elif detector == "ModuleCohesionDetector":
            # Estimate 0.02 modularity improvement
            modified.modularity = min(1.0, modified.modularity + 0.02)

        elif detector in (
            "InappropriateIntimacyDetector",
            "FeatureEnvyDetector",
            "ShotgunSurgeryDetector",
            "DataClumpsDetector",
        ):
            # Estimate 0.5 coupling reduction
            if modified.avg_coupling is not None:
                modified.avg_coupling = max(0.0, modified.avg_coupling - 0.5)

        elif detector == "MiddleManDetector":
            # Removing a middle man reduces bottlenecks
            modified.bottleneck_count = max(0, modified.bottleneck_count - 1)

        return modified

    def _score_structure(self, m: MetricsBreakdown) -> float:
        """Score graph structure metrics (matches engine.py)."""
        modularity_score = m.modularity * 100
        avg_coupling = m.avg_coupling if m.avg_coupling is not None else 0.0
        coupling_score = max(0, 100 - (avg_coupling * 10))
        cycle_penalty = min(50, m.circular_dependencies * 10)
        cycle_score = 100 - cycle_penalty
        bottleneck_penalty = min(30, m.bottleneck_count * 5)
        bottleneck_score = 100 - bottleneck_penalty

        return (modularity_score + coupling_score + cycle_score + bottleneck_score) / 4

    def _score_quality(self, m: MetricsBreakdown) -> float:
        """Score code quality metrics (matches engine.py)."""
        dead_code_score = 100 - (m.dead_code_percentage * 100)
        duplication_score = 100 - (m.duplication_percentage * 100)
        god_class_penalty = min(40, m.god_class_count * 15)
        god_class_score = 100 - god_class_penalty

        return (dead_code_score + duplication_score + god_class_score) / 3

    def _score_architecture(self, m: MetricsBreakdown) -> float:
        """Score architecture health (matches engine.py)."""
        layer_penalty = min(50, m.layer_violations * 5)
        layer_score = 100 - layer_penalty

        boundary_penalty = min(40, m.boundary_violations * 3)
        boundary_score = 100 - boundary_penalty

        # Abstraction: 0.3-0.7 is ideal
        if 0.3 <= m.abstraction_ratio <= 0.7:
            abstraction_score = 100.0
        else:
            distance = min(
                abs(m.abstraction_ratio - 0.3), abs(m.abstraction_ratio - 0.7)
            )
            abstraction_score = max(50, 100 - (distance * 100))

        return (layer_score + boundary_score + abstraction_score) / 3

    def _calculate_overall(
        self,
        structure: float,
        quality: float,
        architecture: float,
    ) -> float:
        """Calculate overall score from category scores."""
        return (
            structure * self.STRUCTURE_WEIGHT
            + quality * self.QUALITY_WEIGHT
            + architecture * self.ARCHITECTURE_WEIGHT
        )

    def _score_to_grade(self, score: float) -> str:
        """Convert numeric score to letter grade."""
        for grade, (min_score, max_score) in GRADES.items():
            if grade == "A":
                if min_score <= score <= max_score:
                    return grade
            else:
                if min_score <= score < max_score:
                    return grade
        return "F"

    def _get_affected_metric(self, detector: str) -> str:
        """Get the metric name affected by a detector."""
        if detector in DETECTOR_METRIC_MAPPING:
            return DETECTOR_METRIC_MAPPING[detector][0]
        return "unknown"

    def _classify_impact(
        self,
        score_delta: float,
        grade_changed: bool,
    ) -> ImpactLevel:
        """Classify the impact level based on score change."""
        if grade_changed or score_delta > 5.0:
            return ImpactLevel.CRITICAL
        elif score_delta > 2.0:
            return ImpactLevel.HIGH
        elif score_delta > 0.5:
            return ImpactLevel.MEDIUM
        elif score_delta > 0.1:
            return ImpactLevel.LOW
        else:
            return ImpactLevel.NEGLIGIBLE


def estimate_fix_impact(
    metrics: MetricsBreakdown,
    finding: Finding,
) -> Dict:
    """Convenience function to estimate impact of fixing a single finding.

    Args:
        metrics: Current codebase metrics
        finding: Finding to estimate impact for

    Returns:
        Dictionary with impact estimation
    """
    calculator = HealthScoreDeltaCalculator()
    delta = calculator.calculate_delta(metrics, finding)
    return delta.to_dict()


def estimate_batch_fix_impact(
    metrics: MetricsBreakdown,
    findings: List[Finding],
) -> Dict:
    """Convenience function to estimate impact of fixing multiple findings.

    Args:
        metrics: Current codebase metrics
        findings: Findings to estimate impact for

    Returns:
        Dictionary with aggregate impact estimation
    """
    calculator = HealthScoreDeltaCalculator()
    delta = calculator.calculate_batch_delta(metrics, findings)
    return delta.to_dict()

File: repotoire-cli/src/detectors/health_delta.rs
//! Health score delta calculator for estimating fix impact
//!
//! This module provides utilities for estimating how resolving a finding
//! would impact the overall health score. This enables before/after
//! comparisons when users review proposed fixes.
//!
//! # Example
//!
//! ```ignore
//! let calculator = HealthScoreDeltaCalculator::new();
//! let delta = calculator.calculate_delta(&current_metrics, &finding);
//! println!("Fixing this would improve score by {:.1} points", delta.score_delta);
//! ```

#![allow(dead_code)] // Module under development - structs/helpers used in tests only

use crate::models::{Finding, Severity};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Impact level classification for health score changes
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ImpactLevel {
    /// >5 points improvement or grade change
    Critical,
    /// 2-5 points improvement
    High,
    /// 0.5-2 points improvement
    Medium,
    /// <0.5 points improvement
    Low,
    /// <0.1 points improvement
    Negligible,
}

impl std::fmt::Display for ImpactLevel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ImpactLevel::Critical => write!(f, "critical"),
            ImpactLevel::High => write!(f, "high"),
            ImpactLevel::Medium => write!(f, "medium"),
            ImpactLevel::Low => write!(f, "low"),
            ImpactLevel::Negligible => write!(f, "negligible"),
        }
    }
}

/// Metrics breakdown for health score calculation
#[derive(Debug, Clone, Default)]
pub struct MetricsBreakdown {
    // Structure metrics
    pub modularity: f64,
    pub avg_coupling: Option<f64>,
    pub circular_dependencies: i32,
    pub bottleneck_count: i32,

    // Quality metrics
    pub dead_code_percentage: f64,
    pub duplication_percentage: f64,
    pub god_class_count: i32,

    // Architecture metrics
    pub layer_violations: i32,
    pub boundary_violations: i32,
    pub abstraction_ratio: f64,

    // Totals for calculations
    pub total_classes: i32,
    pub total_functions: i32,
}

/// Result of a health score delta calculation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthScoreDelta {
    pub before_score: f64,
    pub after_score: f64,
    pub score_delta: f64,
    pub before_grade: String,
    pub after_grade: String,
    pub grade_improved: bool,
    pub structure_delta: f64,
    pub quality_delta: f64,
    pub architecture_delta: f64,
    pub impact_level: ImpactLevel,
    pub affected_metric: String,
    pub finding_id: Option<String>,
    pub finding_severity: Option<Severity>,
}

impl HealthScoreDelta {
    /// Return grade change as string (e.g., "B → A") or None if unchanged
    pub fn grade_change_str(&self) -> Option<String> {
        if self.grade_improved {
            Some(format!("{} → {}", self.before_grade, self.after_grade))
        } else {
            None
        }
    }
}

/// Result of calculating delta for multiple findings
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchHealthScoreDelta {
    pub before_score: f64,
    pub after_score: f64,
    pub score_delta: f64,
    pub before_grade: String,
    pub after_grade: String,
    pub grade_improved: bool,
    pub findings_count: usize,
    pub individual_deltas: Vec<HealthScoreDelta>,
}

/// Mapping from detector names to the metrics they affect
fn detector_metric_mapping() -> HashMap<&'static str, (&'static str, &'static str)> {
    let mut map = HashMap::new();
    // (metric_name, category)
    map.insert(
        "CircularDependencyDetector",
        ("circular_dependencies", "structure"),
    );
    map.insert("GodClassDetector", ("god_class_count", "quality"));
    map.insert("DeadCodeDetector", ("dead_code_percentage", "quality"));
    map.insert("VultureDetector", ("dead_code_percentage", "quality"));
    map.insert(
        "ArchitecturalBottleneckDetector",
        ("bottleneck_count", "structure"),
    );
    map.insert("JscpdDetector", ("duplication_percentage", "quality"));
    map.insert(
        "DuplicateRustDetector",
        ("duplication_percentage", "quality"),
    );
    map.insert("LayerViolationDetector", ("layer_violations", "architecture"));
    map.insert(
        "BoundaryViolationDetector",
        ("boundary_violations", "architecture"),
    );
    map.insert("ModuleCohesionDetector", ("modularity", "structure"));
    map.insert(
        "InappropriateIntimacyDetector",
        ("avg_coupling", "structure"),
    );
    map.insert("FeatureEnvyDetector", ("avg_coupling", "structure"));
    map.insert("ShotgunSurgeryDetector", ("avg_coupling", "structure"));
    map.insert("MiddleManDetector", ("bottleneck_count", "structure"));
    map.insert("DataClumpsDetector", ("avg_coupling", "structure"));
    map
}

/// Grade thresholds
fn score_to_grade(score: f64) -> String {
    if score >= 90.0 {
        "A".to_string()
    } else if score >= 80.0 {
        "B".to_string()
    } else if score >= 70.0 {
        "C".to_string()
    } else if score >= 60.0 {
        "D".to_string()
    } else {
        "F".to_string()
    }
}

/// Calculate health score deltas for individual or batched findings
pub struct HealthScoreDeltaCalculator {
    // Category weights
    structure_weight: f64,
    quality_weight: f64,
    architecture_weight: f64,
}

impl Default for HealthScoreDeltaCalculator {
    fn default() -> Self {
        Self::new()
    }
}

impl HealthScoreDeltaCalculator {
    /// Create a new calculator with default weights
    pub fn new() -> Self {
        Self {
            structure_weight: 0.40,
            quality_weight: 0.30,
            architecture_weight: 0.30,
        }
    }

    /// Create with custom weights
    pub fn with_weights(structure: f64, quality: f64, architecture: f64) -> Self {
        Self {
            structure_weight: structure,
            quality_weight: quality,
            architecture_weight: architecture,
        }
    }

    /// Calculate health score delta for resolving a single finding
    pub fn calculate_delta(&self, metrics: &MetricsBreakdown, finding: &Finding) -> HealthScoreDelta {
        // Calculate current scores
        let current_structure = self.score_structure(metrics);
        let current_quality = self.score_quality(metrics);
        let current_architecture = self.score_architecture(metrics);
        let current_overall =
            self.calculate_overall(current_structure, current_quality, current_architecture);
        let current_grade = score_to_grade(current_overall);

        // Simulate removing the finding's impact
        let modified_metrics = self.remove_finding_impact(metrics, finding);

        // Calculate new scores
        let new_structure = self.score_structure(&modified_metrics);
        let new_quality = self.score_quality(&modified_metrics);
        let new_architecture = self.score_architecture(&modified_metrics);
        let new_overall = self.calculate_overall(new_structure, new_quality, new_architecture);
        let new_grade = score_to_grade(new_overall);

        // Calculate deltas
        let score_delta = new_overall - current_overall;
        let structure_delta = new_structure - current_structure;
        let quality_delta = new_quality - current_quality;
        let architecture_delta = new_architecture - current_architecture;

        // Determine affected metric
        let affected_metric = self.get_affected_metric(&finding.detector);

        // Classify impact level
        let impact_level = self.classify_impact(score_delta, &current_grade != &new_grade);

        HealthScoreDelta {
            before_score: current_overall,
            after_score: new_overall,
            score_delta,
            before_grade: current_grade.clone(),
            after_grade: new_grade.clone(),
            grade_improved: new_grade < current_grade, // A < B < C < D < F
            structure_delta,
            quality_delta,
            architecture_delta,
            impact_level,
            affected_metric,
            finding_id: Some(finding.id.clone()),
            finding_severity: Some(finding.severity),
        }
    }

    /// Calculate health score delta for resolving multiple findings
    pub fn calculate_batch_delta(
        &self,
        metrics: &MetricsBreakdown,
        findings: &[Finding],
    ) -> BatchHealthScoreDelta {
        if findings.is_empty() {
            let current_overall = self.calculate_overall(
                self.score_structure(metrics),
                self.score_quality(metrics),
                self.score_architecture(metrics),
            );
            let current_grade = score_to_grade(current_overall);
            return BatchHealthScoreDelta {
                before_score: current_overall,
                after_score: current_overall,
                score_delta: 0.0,
                before_grade: current_grade.clone(),
                after_grade: current_grade,
                grade_improved: false,
                findings_count: 0,
                individual_deltas: vec![],
            };
        }

        // Calculate current scores
        let current_structure = self.score_structure(metrics);
        let current_quality = self.score_quality(metrics);
        let current_architecture = self.score_architecture(metrics);
        let current_overall =
            self.calculate_overall(current_structure, current_quality, current_architecture);
        let current_grade = score_to_grade(current_overall);

        // Calculate individual deltas
        let individual_deltas: Vec<HealthScoreDelta> = findings
            .iter()
            .map(|f| self.calculate_delta(metrics, f))
            .collect();

        // Simulate removing all findings' impacts
        let mut modified_metrics = metrics.clone();
        for finding in findings {
            modified_metrics = self.remove_finding_impact(&modified_metrics, finding);
        }

        // Calculate new aggregate scores
        let new_structure = self.score_structure(&modified_metrics);
        let new_quality = self.score_quality(&modified_metrics);
        let new_architecture = self.score_architecture(&modified_metrics);
        let new_overall = self.calculate_overall(new_structure, new_quality, new_architecture);
        let new_grade = score_to_grade(new_overall);

        BatchHealthScoreDelta {
            before_score: current_overall,
            after_score: new_overall,
            score_delta: new_overall - current_overall,
            before_grade: current_grade.clone(),
            after_grade: new_grade.clone(),
            grade_improved: new_grade < current_grade,
            findings_count: findings.len(),
            individual_deltas,
        }
    }

    /// Create modified metrics by removing one finding's contribution
    fn remove_finding_impact(
        &self,
        metrics: &MetricsBreakdown,
        finding: &Finding,
    ) -> MetricsBreakdown {
        let mut modified = metrics.clone();
        let detector = &finding.detector;

        // Apply detector-specific adjustments
        if detector == "CircularDependencyDetector" {
            modified.circular_dependencies = (modified.circular_dependencies - 1).max(0);
        } else if detector == "GodClassDetector" {
            modified.god_class_count = (modified.god_class_count - 1).max(0);
        } else if detector == "DeadCodeDetector" || detector == "VultureDetector" {
            // Estimate one dead code item
            let total_nodes = modified.total_classes + modified.total_functions;
            if total_nodes > 0 {
                let per_item_pct = 1.0 / total_nodes as f64;
                modified.dead_code_percentage =
                    (modified.dead_code_percentage - per_item_pct).max(0.0);
            }
        } else if detector == "ArchitecturalBottleneckDetector" {
            modified.bottleneck_count = (modified.bottleneck_count - 1).max(0);
        } else if detector == "JscpdDetector" || detector == "DuplicateRustDetector" {
            // Estimate 0.5% reduction per duplicate finding
            modified.duplication_percentage = (modified.duplication_percentage - 0.005).max(0.0);
        } else if detector == "LayerViolationDetector" {
            modified.layer_violations = (modified.layer_violations - 1).max(0);
        } else if detector == "BoundaryViolationDetector" {
            modified.boundary_violations = (modified.boundary_violations - 1).max(0);
        } else if detector == "ModuleCohesionDetector" {
            // Estimate 0.02 modularity improvement
            modified.modularity = (modified.modularity + 0.02).min(1.0);
        } else if detector == "InappropriateIntimacyDetector"
            || detector == "FeatureEnvyDetector"
            || detector == "ShotgunSurgeryDetector"
            || detector == "DataClumpsDetector"
        {
            // Estimate 0.5 coupling reduction
            if let Some(coupling) = modified.avg_coupling {
                modified.avg_coupling = Some((coupling - 0.5).max(0.0));
            }
        } else if detector == "MiddleManDetector" {
            // Removing a middle man reduces bottlenecks
            modified.bottleneck_count = (modified.bottleneck_count - 1).max(0);
        }

        modified
    }

    /// Score graph structure metrics
    fn score_structure(&self, m: &MetricsBreakdown) -> f64 {
        let modularity_score = m.modularity * 100.0;
        let avg_coupling = m.avg_coupling.unwrap_or(0.0);
        let coupling_score = (100.0 - (avg_coupling * 10.0)).max(0.0);
        let cycle_penalty = (m.circular_dependencies * 10).min(50) as f64;
        let cycle_score = 100.0 - cycle_penalty;
        let bottleneck_penalty = (m.bottleneck_count * 5).min(30) as f64;
        let bottleneck_score = 100.0 - bottleneck_penalty;

        (modularity_score + coupling_score + cycle_score + bottleneck_score) / 4.0
    }

    /// Score code quality metrics
    fn score_quality(&self, m: &MetricsBreakdown) -> f64 {
        let dead_code_score = 100.0 - (m.dead_code_percentage * 100.0);
        let duplication_score = 100.0 - (m.duplication_percentage * 100.0);
        let god_class_penalty = (m.god_class_count * 15).min(40) as f64;
        let god_class_score = 100.0 - god_class_penalty;

        (dead_code_score + duplication_score + god_class_score) / 3.0
    }

    /// Score architecture health
    fn score_architecture(&self, m: &MetricsBreakdown) -> f64 {
        let layer_penalty = (m.layer_violations * 5).min(50) as f64;
        let layer_score = 100.0 - layer_penalty;

        let boundary_penalty = (m.boundary_violations * 3).min(40) as f64;
        let boundary_score = 100.0 - boundary_penalty;

        // Abstraction: 0.3-0.7 is ideal
        let abstraction_score = if (0.3..=0.7).contains(&m.abstraction_ratio) {
            100.0
        } else {
            let distance = (m.abstraction_ratio - 0.3)
                .abs()
                .min((m.abstraction_ratio - 0.7).abs());
            (100.0 - (distance * 100.0)).max(50.0)
        };

        (layer_score + boundary_score + abstraction_score) / 3.0
    }

    /// Calculate overall score from category scores
    fn calculate_overall(&self, structure: f64, quality: f64, architecture: f64) -> f64 {
        structure * self.structure_weight
            + quality * self.quality_weight
            + architecture * self.architecture_weight
    }

    /// Get the metric name affected by a detector
    fn get_affected_metric(&self, detector: &str) -> String {
        detector_metric_mapping()
            .get(detector)
            .map(|(metric, _)| metric.to_string())
            .unwrap_or_else(|| "unknown".to_string())
    }

    /// Classify the impact level based on score change
    fn classify_impact(&self, score_delta: f64, grade_changed: bool) -> ImpactLevel {
        if grade_changed || score_delta > 5.0 {
            ImpactLevel::Critical
        } else if score_delta > 2.0 {
            ImpactLevel::High
        } else if score_delta > 0.5 {
            ImpactLevel::Medium
        } else if score_delta > 0.1 {
            ImpactLevel::Low
        } else {
            ImpactLevel::Negligible
        }
    }
}

/// Convenience function to estimate impact of fixing a single finding
pub fn estimate_fix_impact(metrics: &MetricsBreakdown, finding: &Finding) -> HealthScoreDelta {
    let calculator = HealthScoreDeltaCalculator::new();
    calculator.calculate_delta(metrics, finding)
}

/// Convenience function to estimate impact of fixing multiple findings
pub fn estimate_batch_fix_impact(
    metrics: &MetricsBreakdown,
    findings: &[Finding],
) -> BatchHealthScoreDelta {
    let calculator = HealthScoreDeltaCalculator::new();
    calculator.calculate_batch_delta(metrics, findings)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn create_test_finding(detector: &str) -> Finding {
        Finding {
            id: "test-1".to_string(),
            detector: detector.to_string(),
            severity: Severity::High,
            title: "Test finding".to_string(),
            description: "Test description".to_string(),
            affected_files: vec![PathBuf::from("test.py")],
            line_start: Some(10),
            line_end: Some(20),
            suggested_fix: None,
            estimated_effort: None,
            category: None,
            cwe_id: None,
            why_it_matters: None,
            ..Default::default()
        }
    }

    fn create_test_metrics() -> MetricsBreakdown {
        MetricsBreakdown {
            modularity: 0.7,
            avg_coupling: Some(3.0),
            circular_dependencies: 2,
            bottleneck_count: 3,
            dead_code_percentage: 0.05,
            duplication_percentage: 0.10,
            god_class_count: 2,
            layer_violations: 1,
            boundary_violations: 2,
            abstraction_ratio: 0.5,
            total_classes: 50,
            total_functions: 200,
        }
    }

    #[test]
    fn test_score_to_grade() {
        assert_eq!(score_to_grade(95.0), "A");
        assert_eq!(score_to_grade(85.0), "B");
        assert_eq!(score_to_grade(75.0), "C");
        assert_eq!(score_to_grade(65.0), "D");
        assert_eq!(score_to_grade(55.0), "F");
    }

    #[test]
    fn test_calculate_delta_circular_dep() {
        let calculator = HealthScoreDeltaCalculator::new();
        let metrics = create_test_metrics();
        let finding = create_test_finding("CircularDependencyDetector");

        let delta = calculator.calculate_delta(&metrics, &finding);

        assert!(delta.score_delta > 0.0);
        assert_eq!(delta.affected_metric, "circular_dependencies");
    }

    #[test]
    fn test_calculate_delta_god_class() {
        let calculator = HealthScoreDeltaCalculator::new();
        let metrics = create_test_metrics();
        let finding = create_test_finding("GodClassDetector");

        let delta = calculator.calculate_delta(&metrics, &finding);

        assert!(delta.score_delta > 0.0);
        assert_eq!(delta.affected_metric, "god_class_count");
    }

    #[test]
    fn test_batch_delta_empty() {
        let calculator = HealthScoreDeltaCalculator::new();
        let metrics = create_test_metrics();

        let delta = calculator.calculate_batch_delta(&metrics, &[]);

        assert_eq!(delta.score_delta, 0.0);
        assert_eq!(delta.findings_count, 0);
    }

    #[test]
    fn test_batch_delta_multiple() {
        let calculator = HealthScoreDeltaCalculator::new();
        let metrics = create_test_metrics();
        let findings = vec![
            create_test_finding("GodClassDetector"),
            create_test_finding("CircularDependencyDetector"),
        ];

        let delta = calculator.calculate_batch_delta(&metrics, &findings);

        assert!(delta.score_delta > 0.0);
        assert_eq!(delta.findings_count, 2);
        assert_eq!(delta.individual_deltas.len(), 2);
    }

    #[test]
    fn test_impact_classification() {
        let calculator = HealthScoreDeltaCalculator::new();

        assert_eq!(calculator.classify_impact(6.0, false), ImpactLevel::Critical);
        assert_eq!(calculator.classify_impact(3.0, false), ImpactLevel::High);
        assert_eq!(calculator.classify_impact(1.0, false), ImpactLevel::Medium);
        assert_eq!(calculator.classify_impact(0.3, false), ImpactLevel::Low);
        assert_eq!(
            calculator.classify_impact(0.05, false),
            ImpactLevel::Negligible
        );
        // Grade change always critical
        assert_eq!(calculator.classify_impact(0.1, true), ImpactLevel::Critical);
    }
}
